### PR TITLE
feat(ui): support any yaml reference in link

### LINKED ISSUE #4752: Support external link annotation
# Summary

Like https://github.com/argoproj/argo-cd/pull/4380, we may want to put external links to workflow templates and workflow execution results as well.

# Use Cases

- Check the specifications for workflow.
- Check logs and performance of workflow execution results

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by tczhao**
Resolves #4752

Signed-off-by: Tianchu Zhao <evantczhao@gmail.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

File: ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx [hunk 1]
```diff
@@ -132,6 +132,11 @@ const WorkflowNodeSummary = (props: Props) => {
         });
     }
 
+    const object = props.workflow;
```
Comment (alexec) [inline ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx:135]: I think we should have an object that is like the node status really.

```
{
                            metadata: {
                                namespace: props.workflow.metadata.namespace,
                                name: props.node.id
                            },
                            workflow: workflow,
                            status: {
                                startedAt: props.node.startedAt,
                                finishedAt: props.node.finishedAt
                            }
                        }}
```

That way you can reference the workflow name, and its started and finished at times.

File: ui/src/app/shared/components/links.tsx [hunk 2]
```diff
@@ -5,6 +5,18 @@ import {Link} from '../../../models';
 import {services} from '../services';
 import {Button} from './button';
 
+export const ProcessURL = (url: string, jsonObject: any) => {
+    /* replace ${} from input url with corresponding elements from object
+    return null if element is not found*/
+    return url.replace(/\${[^}]*}/g, x => {
```
Comment (alexec) [inline ui/src/app/shared/components/links.tsx:11]: this is very smart!

File: pkg/apis/workflow/v1alpha1/info.go [hunk 3]
```diff
@@ -8,6 +8,6 @@ type Link struct {
 	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
 	// Either "workflow" or "pod"
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/info.go:9]: minor - but scope can also be some other things

**Review by alexec — CHANGES_REQUESTED** [review]
looks really good so far - one change suggested - dismiss this review when ready to review again

**Review by alexec — APPROVED** [review]

**Comment by zewelor** [comment]
Is there any example how to use this ?

**Comment by tczhao** [comment]
@zewelor 
Hi, you can modify the [links in configmap](https://github.com/argoproj/argo-workflows/blob/master/docs/workflow-controller-configmap.yaml#L41)
using something like _workflow.metadata.annotations.userDefinedKey_ 
the userDefinedKey is from your workflow

note that this feature is only available from master branch at the moment or future 3.1 release

**Comment by zewelor** [comment]
Got it thanks for reply !

**Pull request merged**